### PR TITLE
Begin updating script-library to support arm64

### PR DESCRIPTION
**Pull request by Chuxel**
This PR adds support for arm64 (and in some cases armhf) where possible to scripts in the library-scripts folder (related to #558).  Given this drove a number of changes, I also picked up #511 and Terraform for #571. #840 is also resolved.

There's enough restructuring here that I want to get this in. All scripts work with x86_64.

Remaining issues:
- Some scripts like Homebrew are x86_64 specific since this is not directly supported (e.g. you have to install Ruby yourself from the ground up ... which is slow).
- There is currently an issue I am seeing in Docker that is a seg fault in `gradle-debian.sh` in curl, that doesn't happen for `maven-debian.sh` and `java-debian.sh` that are nearly identical. I will raise a bug on this since it's not blocking to get this in.
- The build system needs to be adapted, but this is tracked as #914.

The `script-library/tests/regression/test.sh` script can now use buildx to test cross-platform builds. More testing is needed on actual arm64 hardware.

### CODE REVIEW

**Comment by Chuxel** [comment]
//cc @joshspicer Since this technically affects the codespaces image.

File: script-library/docker-in-docker-debian.sh [hunk 1]
```diff
@@ -67,30 +67,45 @@ fi
...
+        echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/microsoft-archive-keyring.gpg] https://packages.microsoft.com/repos/microsoft-${ID}-${VERSION_CODENAME}-prod ${VERSION_CODENAME} main" > /etc/apt/sources.list.d/microsoft.list
         apt-get update
         apt-get -y install --no-install-recommends moby-cli moby-buildx moby-engine
     else
-        curl -fsSL https://download.docker.com/linux/$(lsb_release -is | tr '[:upper:]' '[:lower:]')/gpg | (OUT=$(apt-key add - 2>&1) || echo $OUT)
-        echo "deb [arch=amd64] https://download.docker.com/linux/$(lsb_release -is | tr '[:upper:]' '[:lower:]') $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list
+        # Import key safely (new 'signed-by' method rather than deprecated apt-key approach) and install
+        curl -fsSL https://download.docker.com/linux/${ID}/gpg | gpg --dearmor > /usr/share/keyrings/docker-archive-keyring.gpg
+        echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/${ID} ${CODNAME} stable" > /etc/apt/sources.list.d/docker.list
```
Comment (joshspicer) [inline script-library/docker-in-docker-debian.sh:81]: typo: `${CODNAME}`

File: script-library/desktop-lite-debian.sh [hunk 2]
```diff
@@ -94,12 +94,12 @@ apt-get-update-if-needed
 
 # On older Ubuntu, Tilix is in a PPA. On Debian Strech, its in backports
 if [[ -z $(apt-cache --names-only search ^tilix$) ]]; then
-    apt-get install -y --no-install-recommends lsb-release
-    if [ "$(lsb_release -is)" = "Ubuntu" ]; then
+    . /etc/os-release
+    if [ "${ID}" = "ubuntu" ]; then
         apt-get install -y --no-install-recommends apt-transport-https software-properties-common
         add-apt-repository -y ppa:webupd8team/terminix
```
Comment (joshspicer) [inline script-library/desktop-lite-debian.sh:100]: `ppa:webupd8team/terminix` is a legit ppa we want to pull from?

File: script-library/git-from-src-debian.sh [hunk 3]
```diff
@@ -18,13 +19,30 @@ if [ "$(id -u)" -ne 0 ]; then
...
 fi
 
-if [ "${GIT_VERSION}" = "latest" ] || [ "${GIT_VERSION}" = "lts" ] || [ "${GIT_VERSION}" = "current" ]; then
-    RECENT_TAGS=$(curl -sSL -H "Accept: application/vnd.github.v3+json" "https://api.github.com/repos/git/git/tags")
-    GIT_VERSION=$(echo ${RECENT_TAGS} | grep -oE 'name":\s*"v[0-9]+\.[0-9]+\.[0-9]+"' | head -n 1 | sed 's/^name":\s*"v\(.*\)"$/\1/')
-fi
-
 export DEBIAN_FRONTEND=noninteractive
 
+# Source /etc/os-release to get OS info
```
Comment (joshspicer) [inline script-library/git-from-src-debian.sh:24]: Why allow an option to pull from a `ppa` if this script is called `git-from-src-debian.sh`?

File: script-library/go-debian.sh [hunk 4]
```diff
@@ -70,11 +70,20 @@ if [ "${TARGET_GO_VERSION}" = "latest" ] ||  [ "${TARGET_GO_VERSION}" = "current
     TARGET_GO_VERSION=$(curl -sSL "https://golang.org/VERSION?m=text" | sed -n '/^go/s///p' )
 fi
 
+ARCHITECTURE="$(uname -m)"
+case $ARCHITECTURE in
+    x86_64) ARCHITECTURE="amd64";;
+    aarch64 | armv8*) ARCHITECTURE="arm64";;
+    aarch32 | armv7* | armvhf*) ARCHITECTURE="armv6l";;
+    i?86) ARCHITECTURE="386";;
+    *) echo "(!) Architecture unsupported"; exit 1 ;;
```
Comment (joshspicer) [inline script-library/go-debian.sh:79]: Maybe print the `$ARCHITECTURE` in the error `echo` to make it easier for someone (probably us!) to debug if this is reported

File: script-library/powershell-debian.sh [hunk 5]
```diff
@@ -18,19 +18,25 @@ fi
 
 export DEBIAN_FRONTEND=noninteractive
 
-# Install curl, apt-transport-https, lsb-release, or gpg if missing
+# Install curl, apt-transport-https, or gpg if missing
 if ! dpkg -s apt-transport-https curl ca-certificates lsb-release > /dev/null 2>&1 || ! type gpg > /dev/null 2>&1; then
```
Comment (joshspicer) [inline script-library/powershell-debian.sh:22]: `lsb-release` was removed from the comment, but not `if` condition

File: script-library/ruby-debian.sh [hunk 6]
```diff
@@ -93,15 +101,16 @@ else
     # Use a temporary locaiton for gpg keys to avoid polluting image
     export GNUPGHOME="/tmp/rvm-gnupg"
     mkdir -p ${GNUPGHOME}
-    echo "disable-ipv6" >> ${GNUPGHOME}/dirmngr.conf
+    chmod 700 ${GNUPGHOME}
+    echo -e "disable-ipv6\nkeyserver hkps://keys.openpgp.org\nkeyserver hkp://keyserver.ubuntu.com:80\nkeyserver hkp://keyserver.pgp.com" > /tmp/rvm-gnupg/dirmngr.conf
```
Comment (joshspicer) [inline script-library/ruby-debian.sh:105]: This line is tricky to read with the `\n` interwoven

File: script-library/terraform-debian.sh [hunk 7]
```diff
@@ -21,52 +27,121 @@ if [ "$(id -u)" -ne 0 ]; then
 fi
 
 # Install curl, unzip if missing
```
Comment (joshspicer) [inline script-library/terraform-debian.sh:29]: comment is outdated

File: script-library/test/regression/run-scripts.sh [hunk 8]
```diff
@@ -25,37 +39,41 @@ runScript()
...
+    runScript maven "3.6.3 /usr/local/sdkman3 ${USERNAME} false" 
+    runScript node "/usr/local/share/nvm 10 ${USERNAME}"
+    runScript python "3.4.10 /opt/python ${USERNAME} false false"
+    runScript ruby "${USERNAME} false" "2.7.3"
+    runScript rust "/opt/rust/cargo /opt/rust/rustup ${USERNAME} false"
+    runScript terraform "0.8.2 0.12.16"
+    runScript sshd "2223 ${USERNAME} true random"
+    runScript desktop-lite "${USERNAME} changeme false"
+    runScript docker-in-docker "false ${USERNAME} false"
+    if [ "${ARCHITECTURE}" = "amd4" ] || [ "${ARCHITECTURE}" = "x86_64" ] || [ "${ARCHITECTURE}" = "arm64" ] || [ "${ARCHITECTURE}" = "aarch64" ]; then
```
Comment (joshspicer) [inline script-library/test/regression/run-scripts.sh:68]: typo: `amd4`

File: script-library/test/regression/test.sh [hunk 9]
```diff
@@ -1,17 +1,47 @@
...
 set -e
 
-cd "$(cd "$(dirname "$0")" && pwd)/../.."
+cd "$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/../.."
 echo -e  "🧪 Testing image $IMAGE_TO_TEST (${DISTRO}-like)..."
-docker build --build-arg DISTRO=$DISTRO --build-arg IMAGE_TO_TEST=$IMAGE_TO_TEST -f test/regression/Dockerfile .
+
+CURRENT_BUILDERS="$(docker buildx ls)"
+if [[ "${CURRENT_BUILDERS}" != *"vscode-dev-containers"* ]]; then
+    docker buildx create --use --name vscode-dev-containers
```
Comment (joshspicer) [inline script-library/test/regression/test.sh:25]: this line and the else condition look a bit funny to me: One has a `--use` and one has `use`.  Might be fine, not too familiar with buildx and its args

**Review by joshspicer — APPROVED** [review]
A couple small typos to be fixed but otherwise lgtm

File: script-library/desktop-lite-debian.sh (hunk 2 above)
Comment (Chuxel) [reply script-library/desktop-lite-debian.sh:100]: Yep. This part is not new. It only kicks in for Ubuntu 16.04.  18.04+ have it in the main repo.

File: script-library/test/regression/test.sh (hunk 9 above)
Comment (Chuxel) [reply script-library/test/regression/test.sh:25]: Yeah if it already exists, then you would use the `use` command.  The `create --use` command is like running create, then use, but only works if the builder doesn't exist.

File: script-library/docker-in-docker-debian.sh (hunk 1 above)
Comment (Chuxel) [reply script-library/docker-in-docker-debian.sh:81]: Good catch!

File: script-library/git-from-src-debian.sh (hunk 3 above)
Comment (Chuxel) [reply script-library/git-from-src-debian.sh:24]: See #511. There is a very specific circumstance where this works - ubuntu and you choose latest.  Everything else you need to build.  We could rename the script, but that breaks links.  I suppose we could default this to not use the PPA, however.

File: script-library/go-debian.sh (hunk 4 above)
Comment (Chuxel) [reply script-library/go-debian.sh:79]: Good idea.

File: script-library/powershell-debian.sh (hunk 5 above)
Comment (Chuxel) [reply script-library/powershell-debian.sh:22]: Ack - gnupg2 needs to be moved up too.

File: script-library/git-from-src-debian.sh (hunk 3 above)
Comment (Chuxel) [reply script-library/git-from-src-debian.sh:24]: Tweaked default.

File: script-library/docker-in-docker-debian.sh [hunk 10]
```diff
@@ -67,30 +67,45 @@ fi
...
+    if [ "${TARGET_COMPOSE_ARCH}" = "amd64" ]; then
+        TARGET_COMPOSE_ARCH="x86_64"
+    fi
+    if [ "${TARGET_COMPOSE_ARCH}" != "x86_64" ]; then
+        # Use pip to get a version that runns on this architecture
+        if ! dpkg -s python3-pip libffi-dev > /dev/null 2>&1; then
+            apt-get-update-if-needed
+            apt-get -y install --no-install-recommends python3-pip libffi-dev
+        fi
+        pip3 install docker-compose
```
Comment (eitsupi) [inline script-library/docker-in-docker-debian.sh:103]: wouldn't it be better to have `--no-cache-dir` option?

**Pull request merged**